### PR TITLE
[CORDA-2025] target and min platform version (cordapp plugin)

### PR DESCRIPTION
**Pull request by florian-f**
https://r3-cev.atlassian.net/browse/CORDA-2025

### CODE REVIEW

File: cordapp/src/main/kotlin/net/corda/plugins/CordappPlugin.kt [hunk 1]
```diff
@@ -118,6 +122,57 @@ class CordappPlugin : Plugin<Project> {
...
+        val releaseVersion =
+                try {
+                    project.rootProject.ext<String>("corda_release_version")
+                } catch (e: ExtraPropertiesExtension.UnknownPropertyException) {
+                    project.logger.warn(e.message)
+                    null
+                } ?: return null
+        val maybeJar = project.configuration("runtime").filter {
+            "corda-$releaseVersion.jar" in it.toString() || "corda-enterprise-$releaseVersion.jar" in it.toString()
+        }
```
Comment (chrisr3) [inline cordapp/src/main/kotlin/net/corda/plugins/CordappPlugin.kt]: Does code for locating `corda.jar` already exist somewhere? I thought it did...

File: cordapp/src/main/kotlin/net/corda/plugins/CordappPlugin.kt [hunk 2]
```diff
@@ -118,6 +122,57 @@ class CordappPlugin : Plugin<Project> {
...
+        return Pair(targetVersion, minPlatformVersion)
+    }
+
+    /**
+     * Retrieve the Corda platform version from the project's Corda jar (if present)
+     *
+     * @param project The project environment this plugin executes in.
+     * @return The platform version of the projects Corda jar, or null if it cannot be determined.
+     */
+    private fun getPlatformVersion(project: Project): Int? {
```
[the remark this answers is not in the document]
Comment (florian-f) [reply cordapp/src/main/kotlin/net/corda/plugins/CordappPlugin.kt]: That was inspired by `verifyAndGetRuntimeJar`. I talked to @chrisr3 and we could not come up with a way to get to the platform version of the project's Corda dependency other than parsing it out of the Manifest of the corda jar.

File: cordapp/src/main/kotlin/net/corda/plugins/CordappPlugin.kt [hunk 3]
```diff
@@ -118,6 +122,57 @@ class CordappPlugin : Plugin<Project> {
...
+        val releaseVersion =
+                try {
+                    project.rootProject.ext<String>("corda_release_version")
+                } catch (e: ExtraPropertiesExtension.UnknownPropertyException) {
+                    project.logger.warn(e.message)
+                    null
+                } ?: return null
```
[the remark this answers is not in the document]
Comment (florian-f) [reply cordapp/src/main/kotlin/net/corda/plugins/CordappPlugin.kt]: I want to return from the function if this value is null

File: cordapp/src/main/kotlin/net/corda/plugins/CordappPlugin.kt (hunk 1 above)
Comment (florian-f) [reply cordapp/src/main/kotlin/net/corda/plugins/CordappPlugin.kt]: There's something similar in `verifyAndGetRuntimeJar`, but it doesn't do exactly what I want

File: cordapp/src/main/kotlin/net/corda/plugins/CordappPlugin.kt [hunk 4]
```diff
@@ -118,6 +122,57 @@ class CordappPlugin : Plugin<Project> {
...
+                        project.logger.info("Setting target version to ${platformVersion}.")
+                    }
+                    platformVersion
+                } else {
+                    cordapp.info?.targetVersion
+                }
+
+        // If the minimum platform version is not set, default to 1.
+        val minPlatformVersion: Int = cordapp.info?.minPlatformVersion ?: 1
+        check(targetVersion != null) { "Target version was not set and could not be determined from the project's Corda dependency. Please specify the target version of your CorDapp." }
```
[the remark this answers is not in the document]
Comment (florian-f) [reply cordapp/src/main/kotlin/net/corda/plugins/CordappPlugin.kt]: good suggestion, thanks

File: cordapp/src/main/kotlin/net/corda/plugins/CordappPlugin.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (Clintonio) [reply cordapp/src/main/kotlin/net/corda/plugins/CordappPlugin.kt]: val allDeps = project.configurations.flatMap { configuration -> it.allDependencies.map { configuration to it }  }.toSet()
val cordaDeps = allDeps.filter { (it.second.group == "net.corda" || it.second.group == "com.r3.corda") && (it.second.name == "corda") }
Then sort on the "version" field. Pick the lowest. Now you have a pair of configuration to dependency, let's call it "foundCordaDep"
val cordaJar= foundCordaDep.first.files(foundCordaDep.second)

Obviously this code is something I hacked together in Slack. But it's roughly the direction I'd expect.

[the remark this answers is not in the document]
Comment (florian-f) [reply cordapp/src/main/kotlin/net/corda/plugins/CordappPlugin.kt]: Discussed with @Clintonio on slack. Came to the conclusion that it's probably better to not attempt to set the target version automatically and error out instead.

**Pull request merged**